File: 380-insert-delete-getrandom-o1/380-insert-delete-getrandom-o1.rb
class RandomizedSet
    
    attr_accessor :map, :arr, :len
    
    def initialize()
        @map = {}
        @arr = []
        @length = 0
    end


=begin
    :type val: Integer
    :rtype: Boolean
=end
    def insert(val)
        if @map.key? val
            false
        else
            @map[val] = @length
            @arr.push(val)
            @length += 1
            true
        end
    end


=begin
    :type val: Integer
    :rtype: Boolean
=end
    def remove(val)
        if @map.key? val
            idx = @map[val]
            temp = @arr[idx]
            @arr[idx] = arr[-1]
            @arr.pop
            @map[arr[idx]] = idx
            @map.delete val
            @length -= 1
            true
        else
            false 
        end
    end


=begin
    :rtype: Integer
=end
    def get_random()
        @arr[rand(0..@length - 1)]
    end


end

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet.new()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.get_random()

File: medium/380-insert-delete-getrandom-o1/380-insert-delete-getrandom-o1.js

class RandomizedSet {
    constructor() {
        this.map = {};
        this.arr = [];
    }
    
    // idea is to map values to their index in this.arr
    // so can remove and get random in O(1)
    insert(val) {
        if (val in this.map) return false;
        
        this.map[val] = this.arr.length;
        this.arr.push(val);
        return true;
    }
    
    // rather than deleting, pop from end and
    // put popped value in this.add index of map value
    // update index val of map key
    // delete from map
    remove(val) {
        if (val in this.map) {
            const idx = this.map[val],
                  last = this.arr[this.arr.length - 1];
            this.arr[idx] = last;
            this.arr.pop();
            this.map[last] = idx;
            delete this.map[val];
            return true;
        }
        
        return false;
    }
    
    getRandom() {
        return this.arr[Math.floor(Math.random() * this.arr.length)];
    }
};

/** 
 * @param {number} val
 * @return {boolean}
 */


/** 
 * @param {number} val
 * @return {boolean}
 */

/**
 * @return {number}
 */


/** 
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */